data-updates/notebooks/app_ees_device_browser.r

```diff
@@ -0,0 +1,196 @@
+# Databricks notebook source
+# DBTITLE 1,Load dependencies
+source("utils.R")
+
+packages <- c("sparklyr", "DBI", "dplyr", "testthat", "arrow", "stringr")
+
+install_if_needed(packages)
+lapply(packages, library, character.only = TRUE)
+
+ga4_table_name <- "catalog_40_copper_statistics_services.analytics_raw.ees_ga4_device_browser"
+ua_table_name <- "catalog_40_copper_statistics_services.analytics_raw.ees_ua_device_browser"
+scrape_table_name <- "catalog_40_copper_statistics_services.analytics_raw.ees_pub_scrape"
+write_service_table_name <- "catalog_40_copper_statistics_services.analytics_app.ees_service_device_browser"
+write_publication_table_name <- "catalog_40_copper_statistics_services.analytics_app.ees_publication_device_browser"
+
+sc <- spark_connect(method = "databricks")
+
+# COMMAND ----------
+
+# DBTITLE 1,Read in and check table integrity
+# Have to do the group by to avoid duplicates because of the change over day. It's not ideal for avgtimeonpage or bouncerate but given the imact is just on that day I'm not going to worry too much about it for now. 
+
+full_data <- sparklyr::sdf_sql(sc, paste("
+  SELECT 
+   date, pagePath, device, browser, SUM(pageviews) as pageviews, SUM(sessions) as sessions, AVG(avgTimeOnPage) as avgTimeOnPage, AVG(bounceRate) as bounceRate
+  FROM (
+    SELECT 
+    date, pagePath, deviceCategory as device, browser, pageviews, sessions, avgTimeOnPage, bounceRate FROM", ua_table_name, "
+    UNION ALL
+    SELECT 
+    date, pagePath, deviceCategory as device, browser, pageviews, sessions, avgTimeOnPage, bounceRate FROM", ga4_table_name, "
+  ) AS combined_data
+  GROUP BY date, pagePath, device, browser
+  ORDER BY date DESC
+")) %>% collect()
+
+test_that("No duplicate rows", {
+  expect_true(nrow(full_data) == nrow(dplyr::distinct(full_data)))
+})
+
+test_that("Data has no missing values", {
+  expect_false(any(is.na(full_data)))
+})
+
+dates <- create_dates(max(full_data$date))
+
+test_that("There are no missing dates since we started", {
+  expect_equal(
+    setdiff(full_data$date, seq(as.Date(dates$all_time_date), max(dates$latest_date), by = "day")) |>
+      length(),
+    0
+  )
+})
+
+# COMMAND ----------
+
+full_data <- full_data %>%
+  mutate(page_type = case_when(
+    str_detect(pagePath, "/find-statistics/") ~ "Release page",
```

Comment:
Normally I take out /data-guidance and /pre-release-access-list out of here too as they'll currently appear under release page, e.g. in this snippet (from app_ees_release_pageviews)

```
scraped_publications <- sparklyr::sdf_sql(sc, paste("SELECT * FROM", scrape_table_name)) |> collect()

slugs <- unique(scraped_publications$slug)
possible_suffixes <- c("/methodology", "/data-guidance", "/prerelease-access-list")

filtered_data <- sql_data |>
  filter(str_detect(pagePath, "/find-statistics/")) |>
  filter(str_detect(pagePath, paste(slugs, collapse = "|"))) |>
  filter(!str_detect(pagePath, paste(possible_suffixes, collapse = "|")))
```

Reply:
have sorted this now!